### ISSUE TITLE
netlinkrib: permission denied

### ISSUE THREAD

**Issue by NatoBoram** [ipfs/kubo#9513]
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/kubo/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [kubo version](https://dist.ipfs.tech/#kubo) or have an issue updating.

### Installation method

built from source

### Version

```Text
Kubo version: 0.19.0-dev-430a65ad5
Repo version: 13
System version: arm64/android
Golang version: go1.19.4
```

### Config

```json
{
  "API": {
    "HTTPHeaders": {}
  },
  "Addresses": {
    "API": "/ip4/127.0.0.1/tcp/5001",
    "Announce": [],
    "AppendAnnounce": [],
    "Gateway": "/ip4/127.0.0.1/tcp/8080",
    "NoAnnounce": [],
    "Swarm": [
      "/ip4/0.0.0.0/tcp/4001",
      "/ip6/::/tcp/4001",
      "/ip4/0.0.0.0/udp/4001/quic",
      "/ip4/0.0.0.0/udp/4001/quic-v1",
      "/ip4/0.0.0.0/udp/4001/quic-v1/webtransport",
      "/ip6/::/udp/4001/quic",
      "/ip6/::/udp/4001/quic-v1",
      "/ip6/::/udp/4001/quic-v1/webtransport"
    ]
  },
  "AutoNAT": {},
  "Bootstrap": [
    "/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN",
    "/dnsaddr/bootstrap.libp2p.io/p2p/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa",
    "/dnsaddr/bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb",
    "/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",
    "/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
    "/ip4/104.131.131.82/udp/4001/quic/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"
  ],
  "DNS": {
    "Resolvers": {}
  },
  "Datastore": {
    "BloomFilterSize": 0,
    "GCPeriod": "1h",
    "HashOnRead": false,
    "Spec": {
      "mounts": [
        {
          "child": {
            "path": "blocks",
            "shardFunc": "/repo/flatfs/shard/v1/next-to-last/2",
            "sync": true,
            "type": "flatfs"
          },
          "mountpoint": "/blocks",
          "prefix": "flatfs.datastore",
          "type": "measure"
        },
        {
          "child": {
            "compression": "none",
            "path": "datastore",
            "type": "levelds"
          },
          "mountpoint": "/",
          "prefix": "leveldb.datastore",
          "type": "measure"
        }
      ],
      "type": "mount"
    },
    "StorageGCWatermark": 90,
    "StorageMax": "10GB"
  },
  "Discovery": {
    "MDNS": {
      "Enabled": true
    }
  },
  "Experimental": {
    "AcceleratedDHTClient": false,
    "FilestoreEnabled": false,
    "GraphsyncEnabled": false,
    "Libp2pStreamMounting": false,
    "P2pHttpProxy": false,
    "StrategicProviding": false,
    "UrlstoreEnabled": false
  },
  "Gateway": {
    "APICommands": [],
    "HTTPHeaders": {
      "Access-Control-Allow-Headers": [
        "X-Requested-With",
        "Range",
        "User-Agent"
      ],
      "Access-Control-Allow-Methods": [
        "GET"
      ],
      "Access-Control-Allow-Origin": [
        "*"
      ]
    },
    "NoDNSLink": false,
    "NoFetch": false,
    "PathPrefixes": [],
    "PublicGateways": null,
    "RootRedirect": "",
    "Writable": false
  },
  "Identity": {
    "PeerID": "12D3KooWRkhCTGb9ZC33x1QWeWPRNJE43EXbT1zkFuMN7RkvTufo"
  },
  "Internal": {},
  "Ipns": {
    "RecordLifetime": "",
    "RepublishPeriod": "",
    "ResolveCacheSize": 128
  },
  "Migration": {
    "DownloadSources": [],
    "Keep": ""
  },
  "Mounts": {
    "FuseAllowOther": false,
    "IPFS": "/ipfs",
    "IPNS": "/ipns"
  },
  "Peering": {
    "Peers": null
  },
  "Pinning": {
    "RemoteServices": {}
  },
  "Plugins": {
    "Plugins": null
  },
  "Provider": {
    "Strategy": ""
  },
  "Pubsub": {
    "DisableSigning": false,
    "Router": ""
  },
  "Reprovider": {},
  "Routing": {
    "Methods": null,
    "Routers": null
  },
  "Swarm": {
    "AddrFilters": null,
    "ConnMgr": {},
    "DisableBandwidthMetrics": false,
    "DisableNatPortMap": false,
    "RelayClient": {},
    "RelayService": {},
    "ResourceMgr": {},
    "Transports": {
      "Multiplexers": {},
      "Network": {},
      "Security": {}
    }
  }
}
```

### Description

Running `ipfs daemon` results in this error:

```go
Initializing daemon...
Kubo version: 0.19.0-dev-430a65ad5
Repo version: 13
System version: arm64/android
Golang version: go1.19.4

Computing default go-libp2p Resource Manager limits based on:
    - 'Swarm.ResourceMgr.MaxMemory': "939 MB"
    - 'Swarm.ResourceMgr.MaxFileDescriptors': 16384

Applying any user-supplied overrides on top.
Run 'ipfs swarm limit all' to see the resulting limits.

2022-12-18T16:17:31.365Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.388Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.392Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.393Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.394Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
2022-12-18T16:17:31.394Z        ERROR   core    core/builder.go:158     constructing the node: route ip+net: netlinkrib: permission denied

Error: constructing the node (see log for full detail): route ip+net: netlinkrib: permission denied
2022-12-18T16:17:31.395Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
```

**Comment by Jorropo** [comment]
We should add a hint to enable the localisation authorizations in your app.

A better fix would be able to run a transiant node without fetching local addresses but that a bigger refactor.

**Comment by yianding** [comment]
> ### Checklist
> * [x]  This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
> * [x]  I have searched on the [issue tracker](https://github.com/ipfs/kubo/issues?q=is%3Aissue) for my bug.
> * [x]  I am running the latest [kubo version](https://dist.ipfs.tech/#kubo) or have an issue updating.
> 
> ### Installation method
> built from source
> 
> ### Version
> ```yaml
> Kubo version: 0.19.0-dev-430a65ad5
> Repo version: 13
> System version: arm64/android
> Golang version: go1.19.4
> ```
> 
> ### Config
> ```json
> {
>   "API": {
>     "HTTPHeaders": {}
>   },
>   "Addresses": {
>     "API": "/ip4/127.0.0.1/tcp/5001",
>     "Announce": [],
>     "AppendAnnounce": [],
>     "Gateway": "/ip4/127.0.0.1/tcp/8080",
>     "NoAnnounce": [],
>     "Swarm": [
>       "/ip4/0.0.0.0/tcp/4001",
>       "/ip6/::/tcp/4001",
>       "/ip4/0.0.0.0/udp/4001/quic",
>       "/ip4/0.0.0.0/udp/4001/quic-v1",
>       "/ip4/0.0.0.0/udp/4001/quic-v1/webtransport",
>       "/ip6/::/udp/4001/quic",
>       "/ip6/::/udp/4001/quic-v1",
>       "/ip6/::/udp/4001/quic-v1/webtransport"
>     ]
>   },
>   "AutoNAT": {},
>   "Bootstrap": [
>     "/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN",
>     "/dnsaddr/bootstrap.libp2p.io/p2p/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa",
>     "/dnsaddr/bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb",
>     "/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",
>     "/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
>     "/ip4/104.131.131.82/udp/4001/quic/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ"
>   ],
>   "DNS": {
>     "Resolvers": {}
>   },
>   "Datastore": {
>     "BloomFilterSize": 0,
>     "GCPeriod": "1h",
>     "HashOnRead": false,
>     "Spec": {
>       "mounts": [
>         {
>           "child": {
>             "path": "blocks",
>             "shardFunc": "/repo/flatfs/shard/v1/next-to-last/2",
>             "sync": true,
>             "type": "flatfs"
>           },
>           "mountpoint": "/blocks",
>           "prefix": "flatfs.datastore",
>           "type": "measure"
>         },
>         {
>           "child": {
>             "compression": "none",
>             "path": "datastore",
>             "type": "levelds"
>           },
>           "mountpoint": "/",
>           "prefix": "leveldb.datastore",
>           "type": "measure"
>         }
>       ],
>       "type": "mount"
>     },
>     "StorageGCWatermark": 90,
>     "StorageMax": "10GB"
>   },
>   "Discovery": {
>     "MDNS": {
>       "Enabled": true
>     }
>   },
>   "Experimental": {
>     "AcceleratedDHTClient": false,
>     "FilestoreEnabled": false,
>     "GraphsyncEnabled": false,
>     "Libp2pStreamMounting": false,
>     "P2pHttpProxy": false,
>     "StrategicProviding": false,
>     "UrlstoreEnabled": false
>   },
>   "Gateway": {
>     "APICommands": [],
>     "HTTPHeaders": {
>       "Access-Control-Allow-Headers": [
>         "X-Requested-With",
>         "Range",
>         "User-Agent"
>       ],
>       "Access-Control-Allow-Methods": [
>         "GET"
>       ],
>       "Access-Control-Allow-Origin": [
>         "*"
>       ]
>     },
>     "NoDNSLink": false,
>     "NoFetch": false,
>     "PathPrefixes": [],
>     "PublicGateways": null,
>     "RootRedirect": "",
>     "Writable": false
>   },
>   "Identity": {
>     "PeerID": "12D3KooWRkhCTGb9ZC33x1QWeWPRNJE43EXbT1zkFuMN7RkvTufo"
>   },
>   "Internal": {},
>   "Ipns": {
>     "RecordLifetime": "",
>     "RepublishPeriod": "",
>     "ResolveCacheSize": 128
>   },
>   "Migration": {
>     "DownloadSources": [],
>     "Keep": ""
>   },
>   "Mounts": {
>     "FuseAllowOther": false,
>     "IPFS": "/ipfs",
>     "IPNS": "/ipns"
>   },
>   "Peering": {
>     "Peers": null
>   },
>   "Pinning": {
>     "RemoteServices": {}
>   },
>   "Plugins": {
>     "Plugins": null
>   },
>   "Provider": {
>     "Strategy": ""
>   },
>   "Pubsub": {
>     "DisableSigning": false,
>     "Router": ""
>   },
>   "Reprovider": {},
>   "Routing": {
>     "Methods": null,
>     "Routers": null
>   },
>   "Swarm": {
>     "AddrFilters": null,
>     "ConnMgr": {},
>     "DisableBandwidthMetrics": false,
>     "DisableNatPortMap": false,
>     "RelayClient": {},
>     "RelayService": {},
>     "ResourceMgr": {},
>     "Transports": {
>       "Multiplexers": {},
>       "Network": {},
>       "Security": {}
>     }
>   }
> }
> ```
> 
> ### Description
> Running `ipfs daemon` results in this error:
> 
> ```go
> Initializing daemon...
> Kubo version: 0.19.0-dev-430a65ad5
> Repo version: 13
> System version: arm64/android
> Golang version: go1.19.4
> 
> Computing default go-libp2p Resource Manager limits based on:
>     - 'Swarm.ResourceMgr.MaxMemory': "939 MB"
>     - 'Swarm.ResourceMgr.MaxFileDescriptors': 16384
> 
> Applying any user-supplied overrides on top.
> Run 'ipfs swarm limit all' to see the resulting limits.
> 
> 2022-12-18T16:17:31.365Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.388Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.389Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.392Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.393Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.394Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> 2022-12-18T16:17:31.394Z        ERROR   core    core/builder.go:158     constructing the node: route ip+net: netlinkrib: permission denied
> 
> Error: constructing the node (see log for full detail): route ip+net: netlinkrib: permission denied
> 2022-12-18T16:17:31.395Z        ERROR   basichost      basic/basic_host.go:334  failed to resolve local interface addresses     {"error": "route ip+net: netlinkrib: permission denied"}
> ```
> 
> It happens on this hardware:
> 
> ```yaml
> u0_a267@localhost
> -----------------
> OS: Android 13 aarch64
> Host: google Pixel 3
> Kernel: 4.9.332-NekoMiya+
> Uptime: 1 day, 12 hours
> Packages: 146 (dpkg), 1 (pkg)
> Shell: bash 5.2.15
> CPU: Qualcomm SDM845 (8) @ 1.766GHz
> Memory: 2624MiB / 3582MiB
> ```

Can you compile kubo with go 1.19.4?  but it is show me that can not compile kubo with go 1.19 ,only compile with go 1.18 
how do you compile kubo with go 1.19?
thanks

**Comment by NatoBoram** [comment]
@yianding It's been a little while since I've setup Termux again, but it's essentially the same as on Windows/Linux/MacOS.

```sh
pkg install golang git make openssh
git clone git@github.com:ipfs/kubo.git ipfs
cd ipfs
make install
```

**Comment by Khyta** [comment]
I managed to get it 'working' on Android 13 (Termux) by loosely following your steps:

```
pkg install golang
pkg install git
pkg install make
pkg install openssh
git clone https://github.com/ipfs/kubo.git
cd kubo
make install
export PATH=$PATH:/data/data/com.termux/files/home/kubo/cmd/ipfs:/data/data/com.termux/files/home/go/bin
```

And then running `ipfs init`.

I still get the same `netlinkrib: permission denied` but I decided to ignore the error and simply open the Web GUI on `127.0.0.1:5001/ipfs/blabla` which surprisingly worked! I'm less sure about the ability to look up files as a CID I made on my Laptop around 24 hours ago gives a 404 on the Mobile client but is discoverable through ipfs.io and cloudflare-ipfs.

![image](https://github.com/ipfs/kubo/assets/33177286/7134f0bf-c402-46a1-914e-98dec09fd223)

![image](https://github.com/ipfs/kubo/assets/33177286/2c476241-e311-41f4-a4a7-7633f6bdaae0)

**Comment by wlynxg** [comment]
I have found a solution to the issue of route ip+net: netlinkrib: permission denied that occurs with Go on Android 11. I provided an answer to this issue on the Go GitHub page: https://github.com/golang/go/issues/40569#issuecomment-1614781739.

Based on my investigation, I found that your issue is similar to the Go problem mentioned above, and I believe they can be resolved together. I have proposed a solution, which you can find here: https://github.com/wlynxg/anet.

Since I haven't reviewed the source code of "kubo," I am unable to submit a pull request (PR) myself. However, I hope someone else can take my solution and submit a PR to resolve the issue for everyone involved.

**Comment by lidel** [comment]
Triage notes:
- Unsure if still an issue -- does anyone who engaged on this can re-check with Kubo 0.40.0-rc1 or later?
- I am guessing this is till an issue because:
  - the go-netroute v0.4.0 upgrade removes the specific NetlinkRIB call, so the error message may change, but the underlying Android SELinux restriction on netlink sockets likely still causes failures. 
  - The proper fix (letting users plug in anet as a replacement for `net.Interfaces`) is pending in  https://github.com/libp2p/go-libp2p/issues/2986

**Comment by NatoBoram** [comment]
Building `master` throws an error:

```log
plugin/loader/preload.sh > plugin/loader/preload.go
go fmt plugin/loader/preload.go >/dev/null
go install "-trimpath" -ldflags="-X "github.com/ipfs/kubo".CurrentCommit=e931b379e -X "github.com/ipfs/kubo".taggedRelease=" ./cmd/ipfs
# github.com/ipfs/kubo/cmd/ipfs
link: github.com/wlynxg/anet: invalid reference to net.zoneCache
make: *** [cmd/ipfs/Rules.mk:25: cmd/ipfs-install] Error 1
```

But there's a `kubo` package, so we can try that.

```log
Initializing daemon...
Kubo version: 0.41.0-d719fb87b
Repo version: 18
System version: arm64/android
Golang version: go1.26.2
PeerID: 12D3KooWJo2EedfXvwXo2Z4QRob4WUBogPSDmQH3zSp51rvwtDEb
2026-05-04T02:50:07.252Z	ERROR	basichost	basic/addrs_manager.go:776	failed to resolve local interface addresses	{"err": {"Op":"route","Net":"ip+net","Source":null,"Addr":null,"Err":{"Syscall":"netlinkrib","Err":13}}}
2026-05-04T02:50:07.260Z	ERROR	p2pnode	libp2p/discovery.go:44	error starting mdns service: route ip+net: netlinkrib: permission denied
2026-05-04T02:50:08.281Z	ERROR	cmd/ipfs	kubo/daemon.go:982	failed to read listening addresses: route ip+net: netlinkrib: permission denied
Run 'ipfs id' to inspect announced and discovered multiaddrs of this node.
RPC API server listening on /ip4/127.0.0.1/tcp/5001
WebUI: http://127.0.0.1:5001/webui
Gateway server listening on /ip4/127.0.0.1/tcp/8080
Routing V1 API exposed at http://127.0.0.1:8080/routing/v1
Daemon is ready
```

**Comment by lidel** [comment]
Small progress since 2022, but I'm afraid the entire networking side is still broken.

Some triage notes below  to save time someone who wants to poke at Kubo on Android and point LLMs in the right directions (comment below if you had success with any of this):

## Past

In 2022 this was a fatal startup bug: `ipfs daemon` exited with `constructing the node: route ip+net: netlinkrib: permission denied` on Android 11+ and never started.

Cause: Android 11 closed `bind()` on `NETLINK_ROUTE` sockets and blocked `RTM_GETLINK` for non-system apps ([Android docs](https://developer.android.com/training/articles/user-data-ids#mac-11-plus)). Go's stdlib `net.Interfaces()` and `net.InterfaceAddrs()` use a bound netlink socket, so every caller fails with `permission denied`, with or without root.

## Present (2026 Q2)

A go-libp2p change made the failure recoverable. `addrs_manager.go:776` swallows the error and falls back to `[IP4Loopback, IP6Loopback]`. @NatoBoram's `0.41.0` reproduction above shows the daemon now reaches `Daemon is ready`.

What works:

- Daemon boots. RPC API, gateway, and WebUI respond on `127.0.0.1`.

What is still broken (to various degree):

- Address advertisement. The node only ever announces `127.0.0.1` and `::1`, so no remote peer can dial it.
- mDNS / LAN discovery. `libp2p/zeroconf/v2/connection.go` calls `net.Interfaces()`.
- NAT traversal. `go-libp2p/p2p/net/nat/internal/nat/{upnp.go,natpmp.go}` same.
- Source-IP selection for QUIC and reuseport. `libp2p/go-netroute` uses bound netlink. Hurts dialing on multi-homed Android (Wi-Fi + cellular).
- Content fetching. Per @Khyta above, a CID resolvable on `ipfs.io` returns 404 on the mobile node. DHT and Bitswap technically run, but with the address manager pinned to loopback and source-IP selection broken, the node behaves more like an isolated peer than a degraded one. I also suspect other peers will not be able to find you as a provider.

In short: daemon and WebUI work on loopback; as a libp2p peer the Android node is effectively isolated.

NOTE: build also fails on Go 1.23+ (current toolchain `go1.26`):

```
link: github.com/wlynxg/anet: invalid reference to net.zoneCache
```

`wlynxg/anet` (pulled in via `pion/webrtc`) uses `//go:linkname net.zoneCache` to update stdlib's IPv6 zone cache. Go 1.23 disabled cross-package `linkname` by default. Escape hatch: `-ldflags="-checklinkname=0"`. I confirmed locally this builds a valid Android arm64 ELF on Kubo `master`, but the resulting binary still has every limitation above.

## Future

Two paths. They are not mutually exclusive, but each has a different blast radius and different owners.

### Path 1: wait for upstream libp2p fixes

libp2p/go-libp2p#2986 tracks the first item and explicitly defers to go-multiaddr. I elaborated in https://github.com/libp2p/go-libp2p/issues/2986#issuecomment-4371166687 that there is more places to fix:

> IIUC Android 11+ support problem is not limited to `multiformats/go-multiaddr`. Every place below either calls stdlib `net.Interfaces()`/`net.InterfaceAddrs()` or binds a netlink socket. Each has to switch to a "netlink-without-bind path" (`wlynxg/anet` or successor) on `GOOS=android`:
> 
> - `multiformats/go-multiaddr` `net/net.go` (`InterfaceMultiaddrs`). Highest leverage. Called by go-libp2p basichost address resolver and directly by kubo. Without this, the node cannot list its own interface addresses and falls back to loopback. Build-tagged anet swap; PoC works.
> - `libp2p/zeroconf/v2` `connection.go`. Without this, mDNS and LAN peer discovery do not start.
> - `libp2p/go-libp2p/p2p/net/nat/internal/nat/{upnp.go,natpmp.go}`. Without this, UPnP and NAT-PMP discovery do not work behind a home router; inbound connectivity stays zero.
> - `libp2p/go-netroute`. Used by `quicreuse` and `reuseport` for source-IP selection per destination. Needs `RTM_GETROUTE` over an unbound netlink socket; `wlynxg/anet` does not currently expose this. Without it, multi-homed dialing on Android (Wi-Fi + cellular) is unreliable, but I also don't know if any of libp2p transports support this (e.g. `/quic-v1` may not support all features of `quic-go` etc).
> - `wlynxg/anet` (or a successor). Must be Go 1.23+ clean (no `//go:linkname`). Otherwise every patched dep above still ships with `-checklinkname=0`. 
>   - PR wlynxg/anet#9 drops the linkname but reintroduces an IPv6 multicast regression (wlynxg/anet#11).
 
NOTE: Kubo maintainers are not responsible for go-libp2p since the end of September 2025: https://ipshipyard.com/blog/2025-libp2p-maintenance-update/

### Path 2: HTTP-based equivalents inside Kubo get mature enough

I suspect Kubo on Android can route around the platform problems by performing the same p2p operations over HTTP/2. The pieces, in the order they matter:

- **Trustless content retrieval (Bitswap equivalent)** 
  - Spec is at https://specs.ipfs.tech/http-gateways/trustless-gateway/. 
  - Kubo implements the responder side on the gateway port, but exposes only on loopback (`/ip4/127.0.0.1/tcp/8080`) and will trigger network fetch on missing content. 
  - Gap: ship a separate default-enabled trustless server, bound to a non-loopback address, with `Gateway.NoFetch=true`, so other Kubo nodes can fetch from it, but won't trigger retrieval of data that is not present in local blockstore.
- **Content and peer routing** 
  - Foundation already covered by `/routing/v1` (`Routing.Type=autoclient` with `Routing.DelegatedRouters`). 
  - Gap: https://github.com/ipfs/specs/issues/496 
  - Gap: https://specs.ipfs.tech/routing/kad-dht/ HTTP PUT/GET semantics and implementation that exposes DHT server over both libp2p (current) and HTTP (gap)
- **Identify equivalent** 
  - No HTTP spec yet for what `/ipfs/id/1.0.0` does over libp2p (peer ID, supported protocols, advertised addresses). Gap: a small `.well-known` endpoint or its equivalent so HTTP-only nodes can be discovered and their capabilities probed.

Caveats: no AutoNATv2 and no [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md)/[Relay V2](https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md). Ordinary TCP `bind()` works on Android (only `NETLINK_ROUTE` bind is blocked), but inbound reachability through cellular CGNAT and consumer NAT without UPnP is still hard, so a useful HTTP-only Kubo on mobile likely needs a reachability transport like Tor onion services or Snowflake-style routing on top of HTTP-based sockets.

Once those land, a runtime switch (e.g. `Swarm.Libp2pEnabled = false`, TBD) lets the daemon boot without libp2p. No netlink call, no anet, no `-checklinkname=0`. Build-time, a tag can compile out libp2p entirely, similar to how we have one for no-FUSE build. 

If LAN or NAT discovery is wanted later in the HTTP-only build, we can use mDNS and NAT-related packages directly see if we can apply the Sendto-without-bind workaround  to make them work.

-----

This is significant amount of work, time will tell which workstream (1) or (2) lands first. 
Once one of them is ready, we can pick up and revisit this issue.

## Guidance for Android devs/users today

Nothing below produces a fully functional peer. Pick based on what you actually need:

- **Just want to fetch IPFS content on the device.** 
  - Do not run Kubo. Use https://www.npmjs.com/package/@helia/verified-fetch or  https://inbrowser.link/, a PWA gateway client. Read-only access works on mobile; the downside is you cannot publish or pin content this way, need an external pin/storage provider.
- **Want the Kubo binary for local RPC / WebUI / scripting.** 
  - Build with `go build -ldflags="-checklinkname=0" ./cmd/ipfs`. Daemon boots, WebUI loads, RPC works on `127.0.0.1`. Content fetching will be unreliable; mDNS, NAT, and LAN are off; no peer can reach you, but you could build DAGs locally. 
  - You could also fork related libraries and try to create upstream fixes for all touchpoints listed in https://github.com/libp2p/go-libp2p/issues/2986#issuecomment-4371166687
- **Want it to actually and reliably network with current go-libp2p stack.** 
  - Root the device and set SELinux to permissive (per the Termux maintainer's note quoted earlier). That restores `bind()` on `NETLINK_ROUTE`. IIUC there is no other working configuration today.
